jsonrpc: introduce bcsEncoding tag to ease migration to base64

bmwill opened the pull request:
Inorder to ease the transition of SuiEvent.bcs and DynamicFieldInfo.bcsName to being encoded with base64, introduce a tag ("bcsEncoding") to indicate the encoding and gracefully default to base58 when the tag isn't present.

## Description 

Describe the changes or additions included in this PR.

## Test plan 

How did you test the new or updated feature?

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:
- [ ] REST API:

amnn reviewed (approved):
I understand the server-side change, but I'm not fully following how the client knows to deal with it. Here's my understanding of what we want to do. Accepting to unblock assuming this all makes sense (and/or is not an issue):

- Fullnodes from 1.39 should start producing Base64 dynamic field and event BCS, and should tag their responses.
- Clients should look out for the tag, to know that they should decode the response as Base64, or otherwise assume the encoding is Base58.

Most of this logic is handled in `sui-json-rpc-types` where we are introducing types that can be read in one of three formats (implicit Base58, explicit Base58, explicit Base64). Here, I want to confirm that:

- this type is used on both the client side and the server side.
- we don't technically need the "explicit Base58" case.

There was also a change in `sui-types`, to read `DynamicFieldInfo`'s `bcs_name` as Base64 always -- I wasn't following what this was for:

- This seems to go against the idea that the default format will be Base58. (I'm not clear when someone would use this `serde_as` annotation, vs reading using the newly introduced `DynamicFieldInfo` in `sui-json-rpc-types`).
- On that vein, does this mean that existing code needs to change its import for this type to the `sui-json-rpc-types` version?
- I would also have expected a similar change for Events, but I didn't spot one, so that piqued my interest too.

bmwill commented:
> Most of this logic is handled in `sui-json-rpc-types` where we are introducing types that can be read in one of three formats (implicit Base58, explicit Base58, explicit Base64). Here, I want to confirm that:
> 
> * this type is used on both the client side and the server side.

Yes this is used on both server as well as client side rust code. As these types are used both for the server implementation as well as the generated client. This is how jsonrpsee works

> * we don't technically need the "explicit Base58" case.

No we don't, and it wont ever really be populated, but I was trying to cover all the bases. Also in a future release we can, and will, remove the fallback so base58 isn't a thing at all for these types.
 
> There was also a change in `sui-types`, to read `DynamicFieldInfo`'s `bcs_name` as Base64 always -- I wasn't following what this was for:
> 
> * This seems to go against the idea that the default format will be Base58. (I'm not clear when someone would use this `serde_as` annotation, vs reading using the newly introduced `DynamicFieldInfo` in `sui-json-rpc-types`).
> * On that vein, does this mean that existing code needs to change its import for this type to the `sui-json-rpc-types` version?
> * I would also have expected a similar change for Events, but I didn't spot one, so that piqued my interest too.

This is definitely an interesting part of the PR. Essentially the existing `DynamicFieldInfo` type in the sui-types crate was being used both as a type in the JSON-RPC schema as well as a type that we used to store in a rocksdb database. Since that one was used in our DB (serialized as BCS) we can't change the schema without breaking reads from the DB, i introduced a new one that is explicitly a type to use in the JSON-RPC schema so that we'd have the freedom to add the new `bcsEncoding` field. I probably could have done some work to reuse the type (by using a custom deserialize impl) but in order to not have any risk of breaking the bcs schema i just introduced a new type. The existing one now has this annotation

```
#[serde_as(as = "Readable<Base64, _>")]
bcs_name: Vec<u8>,
```

note the `Readable` part where it'll be serialized as Vec<u8> for binary formats and Base64 for readable ones. It would maybe be a good idea to just remove it and always serialize as a binary since we have the jsonrpc type now. Event doesn't have this problem since it already had a JSONRPC specific type.
